### ISSUE TITLE
Running time of compute_fisher becomes longer and longer

### ISSUE THREAD

**Issue by RaptorMai** [seffari/overcoming-catastrophic#2]
I found that the running time of compute_fisher becomes longer and longer. Do you know why this is the case? Thank you so much

**Comment by ariseff** [comment]
Thanks for your question. The Fisher computation is certainly not as efficient as possible. For example, the sampling is not parallelized so this can be quite slow. Not sure if this addresses your issue—perhaps you can elaborate?

**Comment by RaptorMai** [comment]
The phenomenon is the running time of ```compute_fisher``` is linearly increasing as the number of tasks increases. 

I found that the key reason is that for every task, we need to call  ```compute_fisher``` one or a few times. 
Every time when we call  ```compute_fisher```, we need to go through ```for i in range(num_samples):```.
Inside this loop, when we call the following line, ```tf.gradients```  function builds a new backpropagation graph each time it is called, so the reason for the slowdown is that TensorFlow has to parse a new graph on each iteration of the loop. 

```ders = sess.run(tf.gradients(tf.log(probs[0,class_ind]), self.var_list), feed_dict={self.x: imgset[im_ind:im_ind+1]})```

I simple fix is taking tf.gradients out of the loop like this. It accelerates the process a lot. 
```fish_gra = tf.gradients(tf.log(probs[0,class_ind]), self.var_list)```
